server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java

```diff
@@ -176,14 +170,36 @@ void updateSecretSeries(
   }
 
   public int setExpiration(long secretContentId, Instant expiration) {
-    Field<Long> minExpiration = decode()
-        .when(SECRETS_CONTENT.EXPIRY.eq(0L), val(expiration.getEpochSecond()))
-        .otherwise(least(SECRETS_CONTENT.EXPIRY, val(expiration.getEpochSecond())));
+    return dslContext.transactionResult(configuration -> {
+      SecretsContentRecord content = dslContext.select(SECRETS_CONTENT.EXPIRY)
+          .from(SECRETS_CONTENT)
+          .where(SECRETS_CONTENT.ID.eq(secretContentId))
+          .forUpdate()
+          .fetchOneInto(SecretsContentRecord.class);
 
-    return dslContext.update(SECRETS_CONTENT)
-        .set(SECRETS_CONTENT.EXPIRY, minExpiration)
-        .where(SECRETS_CONTENT.ID.eq(secretContentId))
-        .execute();
+      if (content == null) {
+        return 0;
+      }
+
+      Long currentExpiry = content.getExpiry();
+      long epochSeconds = expiration.getEpochSecond();
+
+      Long updatedExpiry = (currentExpiry == null || currentExpiry == 0)
+          ? epochSeconds
+          : Math.min(currentExpiry, epochSeconds);
+
+      int contentsUpdated = dslContext.update(SECRETS_CONTENT)
+          .set(SECRETS_CONTENT.EXPIRY, updatedExpiry)
+          .where(SECRETS_CONTENT.ID.eq(secretContentId))
+          .execute();
+
+      int secretsUpdated = dslContext.update(SECRETS)
+          .set(SECRETS.EXPIRY, updatedExpiry)
+          .where(SECRETS.CURRENT.eq(secretContentId))
+          .execute();
```

Comment:
This is the block which keeps the series expiration and content expiration in sync. However, note that it only applies when the content being updated is the current version of a secret i.e. updating the expiration for non-current content has no effect on the series.